### PR TITLE
Upgrade to atk4/data 3.1

### PR DESCRIPTION
**Pull request by abbadon1334**

### CODE REVIEW

**Comment by abbadon1334** [comment]
if review is ok, can be released

File: src/Validator.php [hunk 1]
```diff
@@ -44,36 +39,21 @@ class Validator
...
     public function __construct(Model $model)
     {
-        $this->model = $model;
-
-        if (!isset($model->validator)) {
-            $model->validator = $this;
-        }
-
-        $model->onHook($model::HOOK_VALIDATE, \Closure::fromCallable([$this, 'validate']));
+        $model->onHook(Model::HOOK_VALIDATE, \Closure::fromCallable([$this, 'validate']));
```
Comment (mvorisek) [inline src/Validator.php:49]: why model prop is no longer needed?

also make sure all you PRs have no text file without end EOL (displayed with 🚫 at the end of the diff)

File: src/Validator.php [hunk 2]
```diff
@@ -44,36 +39,21 @@ class Validator
...
 
     /**
      * Set one rule.
-     *
-     * @param array|string|callable $rules
-     *
-     * @return $this
      */
-    public function rule(string $field, $rules)
+    public function rule(string $field, array $rules): self
```
Comment (mvorisek) [inline src/Validator.php:57]: `$this` return val should be kept

File: tests/BasicTest.php [hunk 3]
```diff
@@ -4,65 +4,81 @@
...
+                ],
+            ],
+        ]);
+    }
 
-        $m->addField('name');
-        $m->addField('age', ['type' => 'number']);
-        $m->addField('type', ['required' => true, 'enum' => ['dog', 'ball']]);
-        $m->addField('tail_length', ['type' => 'number']);
+    protected function getModel(): Model
```
Comment (mvorisek) [inline tests/BasicTest.php]: name it `createModel`

File: tests/BasicTest.php [hunk 4]
```diff
@@ -4,65 +4,81 @@
...
+    {
+        return new ModelTest($this->db);
+    }
 
-        $this->c = new Validator($m);
+    protected function getValidator(Model $model): Validator
```
Comment (mvorisek) [inline tests/BasicTest.php]: also here, `createValidator`, programmer should be aware from the naming instantly what it does/returns

File: src/Validator.php (hunk 1 above)
Comment (abbadon1334) [reply src/Validator.php:49]: i think about this, after i tried it with php 8.1, it gives error and due to the fact that dynamic properties will be removed, so first i removed :
`$model->validator = $this;`
after that i think, why we need the model in the validator, if the model is hooked on validate and the model is passed again in that context?
so i removed even the `$model`
**I know it was public, but i think is really useless to have that unused reference**

File: src/Validator.php (hunk 2 above)
Comment (abbadon1334) [reply src/Validator.php:57]: in the @return docblock for referencing static, no?

File: src/Validator.php (hunk 1 above)
Comment (mvorisek) [reply src/Validator.php:49]: > after that i think, why we need the model in the validator, if the model is hooked on validate and the model is passed again in that context?

thanks, makes sense!

File: src/Validator.php (hunk 2 above)
Comment (mvorisek) [reply src/Validator.php:57]: ? I seen `$this` here, which is much stronger type than `self` or even `static`.

File: src/Validator.php (hunk 1 above)
Comment (abbadon1334) [reply src/Validator.php:49]: the thanks goes to you, for that strict tests that make you rewrite and rethinking everything, is a good pain to feel for the good of the project

File: src/Validator.php (hunk 2 above)
Comment (abbadon1334) [reply src/Validator.php:57]: now i add `@return $this` in docblock and i leave `:self` in the signature, it's ok, if not show me what you suggest

File: src/Validator.php (hunk 1 above)
Comment (mvorisek) [reply src/Validator.php:49]: add https://github.com/atk4/core/pull/329/files#diff-72a46f19b75fd07f055b6fa9a5126e2dc0e3ab69f5ffa5f9f0950c22a691e086R11 trait also to the `Validator` class, then such warnings will be emit also for lower PHP versions. You should do that for every major class excl, the ones that use `DiContainerTrait` as then the trait is already included from there.

File: src/Validator.php (hunk 2 above)
Comment (mvorisek) [reply src/Validator.php:57]: 👍, `$this` is needed to help phpstan to understand/check it is the same instance, also to help the developer to know it is return for method chaining

**Review by mvorisek — APPROVED** [review]

**Pull request merged**